src/aten/sycl/BatchNormKernels.cpp

```diff
@@ -0,0 +1,3965 @@
+bool is_mixed_type(const Tensor& input, const Args&... parameters) {
+  const auto parameter_type = first_type(parameters...);
+  return (
+      (parameter_type != ScalarType::Undefined) &&
+      (parameter_type != input.scalar_type()));
+}
+
+inline bool batch_norm_use_channels_last_kernels(const at::Tensor& self) {
+  return (
+      self.is_contiguous(at::MemoryFormat::ChannelsLast) ||
+      self.is_contiguous(at::MemoryFormat::ChannelsLast3d) ||
+      (self.is_contiguous() && self.strides()[1] == 1));
+}
+
+enum class Impl {
+  Contiguous,
+  ChannelsLast,
+  General,
+};
+
+inline Impl batch_norm_choose_impl(const Tensor& self) {
+  if (!canUse32BitIndexMath(self)) {
+    return Impl::General;
+  }
+
+  if (self.is_contiguous()) {
+    return self.strides()[1] == 1 ? Impl::ChannelsLast : Impl::Contiguous;
+  }
+
+  if (self.is_contiguous(at::MemoryFormat::ChannelsLast)) {
+    return Impl::ChannelsLast;
+  }
+
+  return Impl::General;
+}
+
+inline Impl batch_norm_choose_impl(const Tensor& in1, const Tensor& in2) {
+  auto imp1 = batch_norm_choose_impl(in1);
+  if (imp1 == Impl::General) {
+    return imp1;
+  }
+  auto imp2 = batch_norm_choose_impl(in2);
+  return imp1 == imp2 ? imp1 : Impl::General;
+}
+
+template <
+    typename scalar_t,
+    int64_t dim,
+    template <typename U> class PtrTraits = DefaultPtrTraits,
+    typename index_t = int64_t>
+static GenericPackedTensorAccessor<scalar_t, dim, PtrTraits, index_t>
+get_packed_accessor(const Tensor& t, c10::string_view var_name) {
+  constexpr auto expect_type = c10::CppTypeToScalarType<
+      typename std::remove_const<scalar_t>::type>::value;
+  const auto actual_type = t.scalar_type();
+  TORCH_CHECK(
+      actual_type == expect_type,
+      "Expected ",
+      var_name,
+      " to have type ",
+      expect_type,
+      " but got ",
+      actual_type);
+  return t.generic_packed_accessor<scalar_t, dim, PtrTraits, index_t>();
+}
+
+template <
+    typename scalar_t,
+    int64_t dim,
+    template <typename U> class PtrTraits = DefaultPtrTraits,
+    typename index_t = int64_t>
+static GenericPackedTensorAccessor<scalar_t, dim, PtrTraits, index_t>
+packed_accessor_or_dummy(const Tensor& t, c10::string_view var_name) {
+  if (!t.defined()) {
+    const std::array<index_t, dim> zeros{{0}};
+    return GenericPackedTensorAccessor<scalar_t, dim, PtrTraits, index_t>(
+        nullptr, zeros.data(), zeros.data());
+  }
+  return get_packed_accessor<scalar_t, dim, PtrTraits, index_t>(t, var_name);
+}
+
+struct InvStd {
+  template <typename T>
+  inline T operator()(T var, double epsilon) const {
+    T invstd = 0.0f;
+    if (var != static_cast<T>(0.0f) || epsilon != static_cast<T>(0.0f)) {
+      invstd = static_cast<T>(1.0f) / std::sqrt(var + static_cast<T>(epsilon));
+    }
+    return invstd;
+  }
+};
+
+struct Var {
+  template <typename T>
+  inline T operator()(T var, double epsilon) const {
+    return var;
+  }
+};
+
+int get_max_group_size(int simd = SIMD32) {
```

Comment:
Can we elaborate the purpose/heuristics of the helper? and give it clear comments?